### PR TITLE
Add hidden boolean field to Program

### PR DESCRIPTION
**Pull request by ibrahimahmed443**
Add hidden boolean field to Program model, search index and as a filter.

### CODE REVIEW

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py [hunk 1]
```diff
@@ -226,6 +226,17 @@ def test_filter_by_status(self):
...
+        """ Endpoint should filter programs by their hidden attribute value. """
+        hidden = ProgramFactory(hidden=True)
+        not_hidden = ProgramFactory(hidden=False)
+
+        url = self.list_path + '?hidden=True'
+        self.assert_list_results(url, [hidden], 8)
+
+        url = self.list_path + '?hidden=False'
+        self.assert_list_results(url, [not_hidden], 8)
+
```
Comment (vkaracic) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py:239]: Could you please add a test to verify that the filter field is actually non-nullable. For example if you would filter with `?hidden=None` it shouldn't return any results.

Comment (clintonb) [reply course_discovery/apps/api/v1/tests/test_views/test_programs.py:239]: Frankly, I don't care about any values besides 0, 1, true, and false. As long as the API endpoint returns properly for those, we are fine.

Comment (clintonb) [reply course_discovery/apps/api/v1/tests/test_views/test_programs.py:239]: Please add tests with the field query string set to 0 and 1.

**Review by clintonb — CHANGES_REQUESTED** [review]
Update to your commit message: `Added hidden filter to Programs API list endpoint`.

File: course_discovery/apps/api/filters.py [hunk 2]
```diff
@@ -156,17 +156,15 @@ class Meta:
...
-
-
 class ProgramFilter(FilterSetMixin, django_filters.FilterSet):
     marketable = django_filters.MethodFilter()
     status = django_filters.MultipleChoiceFilter(choices=ProgramStatus.choices)
     type = django_filters.CharFilter(name='type__name', lookup_expr='iexact')
     types = CharListFilter(name='type__slug', lookup_expr='in')
     uuids = UUIDListFilter()
-    hidden = NonNullableBooleanFilter(name='hidden')
+    hidden = django_filters.TypedChoiceFilter(
```
Comment (clintonb) [inline course_discovery/apps/api/filters.py]: Yeah...this is too much of a hack. The first thing you should do is upgrade django-filter to the latest version to determine if the new package corrects issues with `BooleanFilter`. Also, see https://github.com/encode/django-rest-framework/issues/2122 for ideas.

Comment (clintonb) [reply course_discovery/apps/api/filters.py]: See f624a914050aee8748b74220fc27fb51e83f4e8c.

**Comment by ibrahimahmed443** [comment]
@clintonb I fixed some issues with FilterSetMixin. Are we looking good now?

**Comment by clintonb** [comment]
Squash the last four commits.

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -780,7 +780,9 @@ class Program(TimeStampedModel):
         default=False,
         help_text=_('Allow courses in this program to be purchased in a single transaction')
     )
-    hidden = models.BooleanField(default=False, db_index=True)
+    hidden = models.BooleanField(
+        default=False, db_index=True,
+        help_text=_('Hide program on marketing site landing and search pages. This program MAY have a detail page.'))
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Ideally, this should be part of the first commit.

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py [hunk 4]
```diff
@@ -226,6 +226,17 @@ def test_filter_by_status(self):
         url = self.list_path + '?status=active&status=retired'
         self.assert_list_results(url, [retired, active], 8)
 
+    def test_filter_by_hidden(self):
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py:229]: This should be part of the last commit, which deals with API filtering.

File: course_discovery/apps/api/filters.py [hunk 5]
```diff
@@ -161,6 +161,7 @@ class ProgramFilter(FilterSetMixin, django_filters.FilterSet):
     type = django_filters.CharFilter(name='type__name', lookup_expr='iexact')
     types = CharListFilter(name='type__slug', lookup_expr='in')
     uuids = UUIDListFilter()
+    hidden = django_filters.BooleanFilter(name='hidden')
```
Comment (clintonb) [inline course_discovery/apps/api/filters.py]: This belongs in the last commit.

**Review by clintonb — APPROVED** [review]

File: course_discovery/apps/course_metadata/migrations/0056_auto_20170531_0500.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('course_metadata', '0055_program_hidden'),
+    ]
+
+    operations = [
+        migrations.AlterField(
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/migrations/0056_auto_20170531_0500.py]: Merge this with the migration above.

Comment (clintonb) [reply course_discovery/apps/course_metadata/migrations/0056_auto_20170531_0500.py]: All you need to do is move the `help_text` kwarg into the first commit.

**Comment by clintonb** [comment]
@ibrahimahmed443 please remember to delete merged branches.

**Pull request merged**